### PR TITLE
Improve goto-symex documentation

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

File: src/goto-symex/README.md [hunk 1]
```diff
@@ -50,44 +45,85 @@ The \ref bmct creates a goto_symext to symbolically execute the
 goto-program, thereby filling the equation, which can then be passed
 along to the SAT solver.
 
-The symbolic execution state is maintained by goto_symex_statet. This is
-a memory-heavy data structure, so goto_symext creates it on-demand and
-lets it go out-of-scope as soon as possible. However, the process of
-symbolic execution fills out an additional \ref symbol_tablet
+The symbolic execution state is maintained by \ref goto_symex_statet. This
```
Comment (tautschnig) [inline src/goto-symex/README.md]: Maybe we can make this more precise while at it: `goto_symex_statet` is a symbolic representation of the program state at a certain stage during symbolic execution. The state of the symbolic executor is captured in `goto_symext`.

File: src/goto-symex/README.md [hunk 2]
```diff
@@ -50,44 +45,85 @@ The \ref bmct creates a goto_symext to symbolically execute the
...
 function case-switches over the type of the instruction that we're
 currently executing, and calls various other functions appropriate to
 the instruction type, i.e. goto_symext::symex_function_call() if the
 current instruction is a function call, goto_symext::symex_goto() if the
 current instruction is a goto, etc.
 
+\subsection Loop and recursion unwinding
+
+Each backwards goto and recursive call has a separate counter
+(handled by \ref cbmc or another driver program, see the `–unwind` and
```
Comment (tautschnig) [inline src/goto-symex/README.md]: It looks like some editor has wrecked havoc to the double dash? Or is this GitHub being uber-smart?

File: src/goto-symex/README.md [hunk 3]
```diff
@@ -50,44 +45,85 @@ The \ref bmct creates a goto_symext to symbolically execute the
...
+Note the `x$object` fallback candidate, which is known as a *failed symbol* or
+*failed object*, and represents the unknown object pointed to by `x` when
+neither of the candidates (`candidate1` and `candidate2` here) matched as
+expected. This is of course unsound, since `x$object` and `y$object` are always
+distinct, even if `x` and `y` might actually alias, but at least it ensures
+writes to and subsequent reads from `x` are related.
+
+Despite its name, the \ref goto_symext::dereference function is also responsible
+for calling \ref goto_symex_statet::rename, which is responsible for both SSA
+renaming symbols and for applying constant propagation where possible.
```
Comment (tautschnig) [inline src/goto-symex/README.md]: Not quite accurate: renaming is also done in other places.

File: src/goto-symex/README.md [hunk 4]
```diff
@@ -50,44 +45,85 @@ The \ref bmct creates a goto_symext to symbolically execute the
...
 solver. Notably, it _merges_ paths that diverge due to a goto
 instruction, forming a constraint that represents having taken either of
-the two paths; the code for doing this is goto_symext::merge_goto().
-
-CBMC can operate in a different mode when the `--paths` flag is passed
-on the command line. This disables path merging; instead, CBMC executes
-a single path at a time, calling the solver with the equation
+the two paths; the code for doing this is \ref goto_symext::merge_goto.
+
+Symex can operate in a different mode when the `--paths` flag is passed
```
Comment (peterschrammel) [inline src/goto-symex/README.md]: Goto-symex (to avoid confusion with the 'symex' that has been moved out into a separate repo)

**Comment by smowton** [comment]
@tautschnig @peterschrammel updated, ready to go afaik

**Review by tautschnig — APPROVED** [review]

**Pull request merged**